File: 1588.所有奇数长度子数组的和.cpp
/*
 * @lc app=leetcode.cn id=1588 lang=cpp
 *
 * [1588] 所有奇数长度子数组的和
 */

// @lc code=start
class Solution {
public:
    int sumOddLengthSubarrays(vector<int>& arr) {
        int ans=0;
        vector<int> preSum(arr.size()+1,0); //长度arr.size()+1，都是0
        // preSum[ i ] 表示数组 nums 从下标 0 到下标 i - 1 的所有元素之和

        for (int i=0;i<arr.size();++i){
            preSum[i+1]=preSum[i]+arr[i];
        }

        // 从索引为 0 的位置开始遍历出所有，j= 1、3、5 等等奇数长度的子数组
        for(int i=0;i<arr.size();++i){
            for (int j=1;j<=arr.size();j+=2){
                if(i+j-1<arr.size()){
                    ans+=preSum[i+j]-preSum[i];
                }
            }
        }
        return ans;
    }
};
/*
class Solution {
public:
    int sumOddLengthSubarrays(vector<int>& arr) {
        int ans = 0;
        vector<int> preSum(arr.size() + 1, 0);

        for (int i = 0; i < arr.size(); i++) {
            preSum[i + 1] = preSum[i] + arr[i];
        }

        for (int i = 0; i < arr.size(); i++) {
            for (int j = 1; j <= arr.size(); j += 2) {
                if (i + j - 1 < arr.size()) {
                    ans += preSum[i + j] - preSum[i];
                }
            }
        }

        return ans;
    }
};

*/
// @lc code=end

File: 1588.所有奇数长度子数组的和.py
#
# @lc app=leetcode.cn id=1588 lang=python3
#
# [1588] 所有奇数长度子数组的和
#

# @lc code=start
class Solution:
    def sumOddLengthSubarrays(self, arr: List[int]) -> int:
        ans=0
        # presum前缀和
        presum=[0]*(len(arr)+1)

        for i in range(len(arr)):
            presum[i+1]=presum[i]+arr[i]
        
        for i in range(len(arr)):
            # 这里len(arr)+1  需要包含长度len(arr)；cpp版的是<=所以没有+1的问题
            for j in range(1,len(arr)+1,2):
                if i+j-1<len(arr):
                    ans+=presum[i+j]-presum[i]
        return ans
# @lc code=end

